include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h

```diff
@@ -468,22 +500,18 @@ struct __get_sycl_range
         return __range_holder<decltype(rng)>{rng};
     }
 
-    // TODO Add specialization for general case, e.g., permutation_iterator using host
-    // or another fancy iterator.
-    //specialization for permutation_iterator using USM pointer as source
-    template <typename _It, typename _Map, ::std::enable_if_t<!is_hetero_it_v<_It>, int> = 0>
+    //specialization for permutation_iterator using USM pointer or direct pass object as source
+    template <typename _Iter, typename _Map, ::std::enable_if_t<!is_hetero<_Iter>::value
+              && is_passed_directly<_Iter>::value, int> = 0>
```

Comment:
This is getting really "exotic" but what if the there is a `zip_iterator` or `permutation_iterator` that is split between 
a `is_hetero_it` and a `is_passed_directly_it`.  This should be a "valid" input, but will return false for both checks.

I think this is only important for `permutation_iterator(split_iterator, valid_input)` because it special case

To be more explicit:
`permutation_iterator(zip_iter(hetero_it,passed_directly_it), valid_input)`
`permutation_iterator(zip_iter(passed_directly_it, hetero_it), valid_input)`
`permutation_iterator(permutation_iterator(passed_directly_it, hetero_it), valid_input)`
`permutation_iterator(permutation_iterator(hetero_it, passed_directly_it), valid_input)`

I think none of these will find an overload that fits it.  

We could create a third trait `is_hetero_or_passed_directly`, which has the appropriate logic for these split iterators.  However, I don't know if it is possible to support such a case with another `permutation_iterator` overload.  I think you run into the problem of not knowing the size of the range of the split iterator that is needed, which is problematic when it is split.

I'd recommend just adding this as a known limitation perhaps.

Reply:
Yes, it seems there is no way to support  `permutation_iterator(non_passed_directly, valid_input)`. 
Regarding  "adding  known limitation perhaps" - in that case we have to also explain what is "passed directly"... But I am not sire that we want to do that.